// @stoplight/spectral/dist/formats/json-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isJSONSchemaDraft2019_09 = exports.isJSONSchemaDraft7 = exports.isJSONSchemaDraft6 = exports.isJSONSchemaDraft4 = exports.isJSONSchemaLoose = exports.isJSONSchema = void 0;
const lodash_1 = require("lodash");
const KNOWN_JSON_SCHEMA_TYPES = ['array', 'boolean', 'integer', 'null', 'number', 'object', 'string'];
const KNOWN_JSON_SCHEMA_COMBINERS = ['allOf', 'oneOf', 'anyOf'];
const $SCHEMA_DRAFT4_REGEX = /^https?:\/\/json-schema.org\/draft-04\/(?:hyper-)?schema#?$/;
const $SCHEMA_DRAFT6_REGEX = /^https?:\/\/json-schema.org\/draft-06\/(?:hyper-)?schema#?$/;
const $SCHEMA_DRAFT7_REGEX = /^https?:\/\/json-schema.org\/draft-07\/(?:hyper-)?schema#?$/;
const $SCHEMA_DRAFT_2019_09_REGEX = /^https?:\/\/json-schema.org\/draft\/2019-09\/(?:hyper-)?schema#?$/;
const hasValidJSONSchemaType = (document) => 'type' in document && typeof document.type === 'string' && KNOWN_JSON_SCHEMA_TYPES.includes(document.type);
const hasValidJSONSchemaCombiner = (document) => KNOWN_JSON_SCHEMA_COMBINERS.some(combiner => combiner in document && lodash_1.isObject(document[combiner]));
exports.isJSONSchema = (document) => lodash_1.isObject(document) &&
    '$schema' in document &&
    String(document.$schema).includes('//json-schema.org/');
exports.isJSONSchemaLoose = (document) => lodash_1.isObject(document) &&
    (exports.isJSONSchema(document) ||
        hasValidJSONSchemaType(document) ||
        hasValidJSONSchemaCombiner(document) ||
        'not' in document);
exports.isJSONSchemaDraft4 = (document) => exports.isJSONSchema(document) && $SCHEMA_DRAFT4_REGEX.test(document.$schema);
exports.isJSONSchemaDraft6 = (document) => exports.isJSONSchema(document) && $SCHEMA_DRAFT6_REGEX.test(document.$schema);
exports.isJSONSchemaDraft7 = (document) => exports.isJSONSchema(document) && $SCHEMA_DRAFT7_REGEX.test(document.$schema);
exports.isJSONSchemaDraft2019_09 = (document) => exports.isJSONSchema(document) && $SCHEMA_DRAFT_2019_09_REGEX.test(document.$schema);
