// @stoplight/spectral/dist/cli/services/linter/utils/listFiles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listFiles = void 0;
const path_1 = require("@stoplight/path");
const fg = require("fast-glob");
const GLOB_OPTIONS = {
    absolute: true,
    dot: true,
};
async function match(pattern) {
    return (await fg(pattern, GLOB_OPTIONS)).map(path_1.normalize);
}
async function listFiles(pattens, ignoreUnmatchedGlobs) {
    const { files, fileDescriptors, urls } = pattens.reduce((group, pattern) => {
        if (typeof pattern === 'number') {
            group.fileDescriptors.push(pattern);
        }
        else if (!/^https?:\/\//.test(pattern)) {
            group.files.push(pattern.replace(/\\/g, '/'));
        }
        else {
            group.urls.push(pattern);
        }
        return group;
    }, {
        files: [],
        urls: [],
        fileDescriptors: [],
    });
    const filesFound = [];
    const fileSearchWithoutResult = [];
    if (ignoreUnmatchedGlobs) {
        filesFound.push(...(await match(files)));
    }
    else {
        await Promise.all(files.map(async (pattern) => {
            const resultFg = await match(pattern);
            if (resultFg.length === 0) {
                fileSearchWithoutResult.push(pattern);
            }
            filesFound.push(...resultFg);
        }));
    }
    return [[...urls, ...fileDescriptors, ...filesFound], fileSearchWithoutResult];
}
exports.listFiles = listFiles;
