// @stoplight/lifecycle/emitter/emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Emitter = require("wolfy87-eventemitter");
const disposable_1 = require("../disposable");
class EventEmitter {
    constructor() {
        this._emitter = new Emitter();
    }
    on(type, listener) {
        const wrappedListener = (...args) => {
            try {
                listener(...args);
            }
            catch (ex) {
                console.error(ex);
            }
        };
        this._emitter.on(String(type), wrappedListener);
        return disposable_1.createDisposable(() => {
            this._emitter.off(String(type), wrappedListener);
        });
    }
    emit(type, ...args) {
        this._emitter.trigger(String(type), args);
    }
    get hasListeners() {
        const eventsToListeners = this._emitter.getListeners(/.*/);
        for (const ev in eventsToListeners) {
            if (!{}.hasOwnProperty.call(eventsToListeners, ev))
                continue;
            const l = eventsToListeners[ev];
            if (l !== undefined && l.length > 0) {
                return true;
            }
        }
        return false;
    }
    dispose() {
        this._emitter.removeAllListeners();
    }
    createEmitGroup() {
        const notifier = this;
        const eventQueue = [];
        let flushed = false;
        return {
            get queueCount() {
                return eventQueue.length;
            },
            emit(event, ...args) {
                if (flushed) {
                    notifier.emit(event, ...args);
                }
                else {
                    eventQueue.push([event, args]);
                }
            },
            flush() {
                for (const [event, args] of eventQueue) {
                    try {
                        notifier.emit(event, ...args);
                    }
                    catch (e) {
                    }
                }
                this.reset();
                flushed = true;
            },
            reset() {
                eventQueue.length = 0;
                flushed = false;
            },
        };
    }
}
exports.EventEmitter = EventEmitter;
