// @stoplight/spectral/dist/functions/schemaPath.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemaPath = void 0;
const jsonpath_plus_1 = require("jsonpath-plus");
const getLintTargets_1 = require("../runner/utils/getLintTargets");
const schema_1 = require("./schema");
exports.schemaPath = (targetVal, opts, paths, otherValues) => {
    var _a;
    const relevantItems = getLintTargets_1.getLintTargets(targetVal, opts.field);
    const schemaObject = jsonpath_plus_1.JSONPath({ path: opts.schemaPath, json: targetVal })[0];
    const results = [];
    for (const relevantItem of relevantItems) {
        const result = schema_1.schema(relevantItem.value, {
            schema: schemaObject,
            oasVersion: opts.oasVersion,
            allErrors: opts.allErrors,
        }, {
            given: paths.given,
            target: [...((_a = paths.target) !== null && _a !== void 0 ? _a : paths.given), ...relevantItem.path],
        }, otherValues);
        if (Array.isArray(result)) {
            results.push(...result);
        }
    }
    return results;
};
